### PR TITLE
F compression function implementation for Blake2b

### PR DESCRIPTION
**Pull request by pdyraga**
Closes: #1 

`F` is a compression function for Blake2. It takes as an argument the
state vector `h`, message block vector `mb`, offset counter `t`,
final block indicator flag `f`, and the number of rounds `rounds` to execute.
The state vector is modified in-place by the function. 

The code has been borrowed from `golang/crypto/blake2b` and adjusted so
that this function can be safely exposed as public. Parameters have
been also adjusted to match those in [RFC 7693](https://tools.ietf.org/html/rfc7693)

Test vectors were generated from those in `golang/crypto/blake2b`

### CODE REVIEW

File: f.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	{0, 2, 4, 6, 1, 3, 5, 7, 8, 10, 12, 14, 9, 11, 13, 15}, // equal to the first
+	{14, 4, 9, 13, 10, 8, 15, 6, 1, 0, 11, 5, 12, 2, 7, 3}, // equal to the second
+}
+
+// F is a compression function for Blake2. It takes as an argument the state
+// vector `h`, message block vector `blocks` 2-bit offset counter `t`, final
+// block indicator flag `f`, and number of rounds `rounds`. The state vector
+// is modified by the function. Number of rounds can be anything between 0 and
+// 12 (inclusive).
+func F(h *[8]uint64, blocks []byte, t *[2]uint64, f uint64, rounds uint8) error {
```
Comment (pdyraga) [inline f.go]: Is [0,12] rounds enough for all our use cases?

File: f_test.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+			F(&h, mHex, &c, 0, 12)
+
+			if !reflect.DeepEqual(test.hOut, h) {
+				t.Errorf("Unexpected result\nExpected: [%v]\nActual:   [%v]\n", test.hOut, h)
+			}
+		})
+	}
+}
+
+var testVectors = []struct {
```
Comment (pdyraga) [inline f_test.go]: We need some better test vectors, ideally something from Zcash.

File: f.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/bits"
+)
+
+const (
+	// BlockSize is the blocksize of BLAKE2 in bytes.
+	BlockSize = 128
```
Comment (pdyraga) [inline f.go:16]: `BlockSize` should be always `128` as far as I understand. Is that correct?

**Comment by mhluongo** [comment]
The Zcash team has confirmed we're good on rounds and blocksize. I gave a shot at generating test vectors with `pyblake2` but it doesn't expose `F` - maybe we should pull in the rest of the blake2 implementation and generate test vectors for that? If a robust blake2 suite works we can be fairly confident in `F`

File: f.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+		v13 ^= t1
+		v14 ^= f
+
+		for j := range m {
+			m[j] = binary.LittleEndian.Uint64(blocks[i:])
+			i += 8
+		}
+
+		for j := 0; j < int(rounds); j++ {
+			s := &(precomputed[j])
```
Comment (mhluongo) [inline f.go]: We can support up to `2 ^ 8 - 1` rounds if we replace this `j` with `j mod 10` and get rid of the last two values in the `precompute` array. It'll be at a slight performance expense but 🤷‍♂️

**Review by mhluongo — COMMENTED** [review]
Interested in your thoughts on raising the round limit to make this more general @pdyraga

**Comment by mhluongo** [comment]
Looks like this will close #1 and #3 as well

File: f.go (hunk 4 above)
Comment (pdyraga) [reply f.go]: Done!

**Comment by pdyraga** [comment]
@mhluongo All comments addressed, I am happy with what's here. It's ready for your review 👀.

**Comment by pdyraga** [comment]
#3 will be covered in go-ethereum PR - each contract has benchmarks there.

File: f.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
+// Copyright 2016 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Modified by KEEP SEZC to expose F compression function.
```
Comment (mhluongo) [inline f.go:5]: Going to touch this up after this PR

**Review by mhluongo — APPROVED** [review]
:rocket: :rocket: :rocket:

File: f.go (hunk 5 above)
Comment (pdyraga) [reply f.go:5]: Yeah, I had no idea what to put here. Worth double-checking if we do not have to alter `LICENSE` as well.

**Comment by pdyraga** [comment]
:tada:

**Comment by carterpy** [comment]
hey @pdyraga here are those questions: 

@pdyraga, I've been playing with your f function in go and it's great and you've done a fantastic job here - massive props. I have two questions but please do not answer the stupid ones:

    In F.go, should f return h?
    Should the package be the same name as the folder it's in?

**Pull request merged**